File: jvm/src/sap/FirstUnique.java
package sap;

import java.util.HashMap;
import java.util.Map;

public class FirstUnique {

    /*
        input: supercalifragilisticexpialidocious
        output: f
    */

    public Character firstUnique(String input) {
        // Char -> whether is unique or not
        final Map<Character, Boolean> m = new HashMap<>();

        for (char c: input.toCharArray()) {
            if(m.getOrDefault(c, false)) {
                m.put(c, false);
            } else m.put(c, true);
        }

        for (char c: input.toCharArray()) {
            if(m.getOrDefault(c, false)) return c;
        }

        return null;
    }
}

File: scala/src/main/scala/sap/FirstUnique.scala
package sap

import scala.collection.immutable.HashMap

object FirstUnique {
  def main(args: Array[String]): Unit = {
    /*
        input: supercalifragilisticexpialidocious
        output: f
    */
    assert(firstUnique("supercalifragilisticexpialidocious") == 'f')
  }

  def firstUnique(s: String): Char = {
    val cs = s.toList
    val m = getOccurrence(cs)

    @annotation.tailrec
    def helper(cs: List[Char]): Char = {
      cs match {
        case Nil => ' '
        case h :: t => if (m(h) == 1) h else helper(t)
      }
    }

    helper(cs)
  }

  def getOccurrence(cs: List[Char]): Map[Char, Int] = cs.groupBy(identity).map {
    case (k, v) => k -> v.length
  }
}
